substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSubsectionBuilder.java

```diff
@@ -126,40 +146,118 @@ private void buildFunction(CompiledMethodEntry compiledEntry) {
 
-        /* S_END add end record. */
-        addSymbolRecord(new CVSymbolSubrecord.CVSymbolEndRecord(cvDebugInfo));
+        addParameters(compiledEntry, varRangeMap);
+        /* In the future: addLocals(compiledEntry, varRangeMap); */
+
+        /* S_PROC_ID_END add end record. */
+        addSymbolRecord(new CVSymbolSubrecord.CVSymbolProcIdEndRecord(cvDebugInfo));
 
         /* Add line number records. */
         addLineNumberRecords(compiledEntry);
     }
 
+    void addParameters(CompiledMethodEntry primaryEntry, Map<LocalEntry, List<Range>> varRangeMap) {
+        final Range primaryRange = primaryEntry.primary();
+        /* The name as exposed to the linker. */
+        final String externalName = primaryRange.getSymbolName();
+        final MethodEntry method = primaryRange.isPrimary() ? primaryRange.getMethodEntry() : primaryRange.getCallees().getFirst().getMethodEntry();
+
+        /* define function parameters */
+        assert Modifier.isStatic(method.getModifiers()) || !method.getParams().isEmpty();
+        for (LocalEntry paramInfo : method.getParams()) {
+            final TypeEntry paramType = paramInfo.type();
+            final int typeIndex = method.isThisParam(paramInfo)
+                            ? cvDebugInfo.getCVTypeSection().getIndexForPointer(paramType)
+                            : cvDebugInfo.getCVTypeSection().addTypeRecords(paramType).getSequenceNumber();
+            emitLocal(paramInfo, varRangeMap, paramInfo.name(), paramType, typeIndex, true, externalName, primaryRange);
+        }
+    }
+
+    private static int infoTypeToInt(LocalValueEntry info) {
+        return switch (info) {
+            case RegisterValueEntry p -> p.regIndex();
+            case StackValueEntry p -> -p.stackSlot();
+            default -> 0;
+        };
+    }
+
+    void emitLocal(LocalEntry info, Map<LocalEntry, List<Range>> varRangeMap, String name, TypeEntry typeEntry, int typeIndex, boolean isParam,
+                    String procName, Range range) {
+        int flags = isParam ? S_LOCAL_FLAGS_IS_PARAM : 0;
+        List<Range> ranges = varRangeMap.get(info);
+        addSymbolRecord(new CVSymbolSubrecord.CVSymbolLocalRecord(cvDebugInfo, name, typeIndex, flags));
+        long currentHigh = Integer.MIN_VALUE;
+        int registerOrSlot = 0; /* -slot or +register or 0=unknown */
+        CVSymbolSubrecord.CVSymbolDefRangeBase currentRecord = null;
+        for (Range subrange : ranges) {
+            LocalValueEntry value = subrange.lookupValue(info);
+            if (value != null) {
+                if (subrange.getLo() == currentHigh && registerOrSlot == infoTypeToInt(value)) {
+                    /* if we can, merge records */
+                    currentHigh = subrange.getHi();
+                    if (currentRecord != null) {
+                        currentRecord.length = (short) (currentHigh - currentRecord.procOffset - range.getLo());
```

Comment:
I don’t think we should assume the combined length fits in a `short`. Could we check before merging and skip otherwise?